### DISCUSSION TITLE
Tasmota32-Display on Esp32C3

### DISCUSSION THREAD

**Discussion by derFliXX** [arendst/Tasmota#20695] [Q&A]
Hi guys...
is it possible to use the tasmota-display version on a esp32C3?
I'm trying to flash something like this for hours but i can't manage it....
Thank you for your help :)

**Comment by sfromis** [comment]
For ESP32-C3, you can **only** use binaries there the name starts with `tasmota32c3-`, including:
https://github.com/tasmota/install/blob/firmware/firmware/unofficial/tasmota32c3-display.factory.bin

Depending on what display you use (if not one of the smallest), I'd recommend LVGL/HASPmota providing better features, but with ESP32-C3 having no PSRAM, you'll quickly run out of RAM if doing more than simple things.
https://github.com/tasmota/install/blob/firmware/firmware/unofficial/tasmota32c3-lvgl.factory.bin

**Reply by jpmorrison** [reply]
The unofficial tasmota32c3-lvgl.bin won't load/upgrade either through file upload or ota. 
Is there a proper ota url for these unofficial blobs or a way to combine them? 

I didn't know the C3 would have RAM issues. I just want to get this ESP32-2424S012 running. 
The LVGL demo looks fine. With nothing enabled running tasmota32c3.bin I have 172.5 KB free RAM.

**Reply by barbudor** [reply]
Have you looked at this discussion ? https://github.com/arendst/Tasmota/discussions/19487

If you can't upgrade, it may because you are not using the safeboot partition mode ?
See the bottom of the INFORMATION page from the web UI

**Reply by sfromis** [reply]
Those `.factory` links above are not for use with OTA, neither via URL nor via file upload, as the `.factory` binaries are **only** for wired initial flashing.

Also, the URL leads to a page where you can find the binary, is not directly usable in OTA via http.

**Reply by jpmorrison** [reply]
I have read through that thread. 

I do have the safeboot partition. I was running tasmota32c3.bin, uploaded a new image -  it failed. On reboot I'm in safeboot. None of the esp32c3-lvgl work, either through ota url or file upload.  Trying esptool with the 32c3-lvgl-factory.bin gave me a boot loop.

**Reply by sfromis** [reply]
Did you read what I wrote about using `.factory` binaries?

**Comment by derFliXX** [comment]
Ah nice thank you... it worked now to flash the tasmota32c3-display version.
I have a Waveshare 7.3inch ePaper display... I connected it via SPI but it seems like the ESP doesnt register it.
In the console i see no indicators for recognition and when i try to test it with "DisplayText" i just get an error... do you maybe have an idea, what I'm doing wrong?

**Reply by sfromis** [reply]
In general, it is because you did not configure it; just connecting it does not mean that Tasmota can "see" it. Such configuration needs to be compatible with the specific display type, or you may need to adapt. Just configuring SPI pins is not sufficient, but Tasmota *does* have pin configs for `EPaper29 CS` and `EPaper42 CS`, which should be included with `-display` builds. I'm "unconvinced" if/that this indicates support for your particular display.

Of possible interest: https://tasmota.github.io/docs/Displays/#waveshare-display-drivers

**Reply by derFliXX** [reply]
What do you mean with configuring? I've also uploaded a "display.ini" file with some configuring stuff (i also modified the height and width parameters as said in the section with the waveshare display drivers).
And why is it possible then to config any pins as SPI pins if I need a special configuration? And I think the display is from the same producer - just with another size - i think it shouldnt be a problem with support bc it says that the drivers should also support other Waveshare models or am I wrong?

**Reply by sfromis** [reply]
Configuring is not just SPI pins, but the mentioned two CS pins hint at activating different types, not simply Waveshare in general. Also, same producer does not automatically mean that differences is only about height/width. To be clear, I'm not making a statement about how different or similar your *specific* display is, merely that while you cannot "just assume" strong commonality, it still *might* be so. Know your hardware.

Even assuming that one of the two Waveshare variants (with or without partial update) is workable, you still need to know which, and configure accordingly.

When it comes to using `display.ini`, that's a different way of configuring when you have an unused pin as `Option A 3` to use uDisplay. The 3 different e-paper configs from https://github.com/arendst/Tasmota/tree/development/tasmota/displaydesc are rather different, and without experience with any of the Waveshare e-paper displays, I cannot make specific suggestions.

When I wrote "In general, it is because you did not configure it", that is due to lots of commands (including `DisplayText`) only being recognized as valid commands when a driver supporting the command has been successfully initialized. Hence, a message like command unknown indicates that Tasmota did was not able to initialize the driver. Of course, as you merely said "i just get an error", you did not provide any info to make more specific guesses.

Configuring SPI pins are to have Tasmota know what pins are used for the SPI bus, it does not tell what device is connected, and generally Tasmota cannot automatically detect SPI devices. Hence further config is needed, like what you can do in `display.ini` if you use uDisplay. And it is possible that more details than just width/height could need tweaking.

**Comment by Jason2866** [comment]
ePaper Displays are the most sophisticated ones to setup. For the one you bought 7.3" does not exists a ready made display.ini. And you are very wrong, every ePaper Display display.ini differs a lot.
